### ISSUE TITLE
SignalR crazy idea

### ISSUE THREAD

**Issue by dazinator** [MarimerLLC/cslaforum#279]
SignalR is basically pub sub, over sockets (if available) between client browsers and a "hub" which is a component running within your server (asp.net) application. Microsoft tout it quite a bit, and it's great if you want to write a chat application.

I have a crazy idea, but am wondering if someone has beaten me to it, or if the topic has come up before.

The idea would be to hold an instance of a business object in memory on the server, then use SignalR to allow the client browser to subscribe to property change notifications, and also to update properties of the business object, via appropriate signalr javascript api's.

I'd call it Signal-CSLA

Basically this would be a bit like remoting, but using signal r as the transport mechanism between the browser and the server side component.

Anyone done something like this? Remember silverlight died right?

**Comment by rockfordlhotka** [comment]
I'm not aware of anyone doing this, but it sounds quite interesting. 

The only real issue I see is that you end up with two points of failure: the client and the server. As opposed to a smart client app where you have one point of failure: the client.

**Comment by dazinator** [comment]
Thanks for the feedback Rocky. I'll try and spike something out at some point just to see what it looks like.

**Comment by dazinator** [comment]
Will leave open for a short while just in case anyone else has any input

**Comment by dazinator** [comment]
I think the thing that appeals to me the most with this, is that there would be no need to try and duplicate any of the business logic / rules to run on the client.. which means less javascript. As much as I have embraced modern javascript frameworks, less is usually more :-)

**Comment by rockfordlhotka** [comment]
Are you thinking that the server would run in ASP.NET, or in a Windows Service?

The risk running in ASP.NET is that you don't control the AppDomain lifecycle, so you may need to consider setting up some sort of suspend/resume model to serialize and then restore your business objects in case the AppDomain is recycled. I don't know how/if you can do that, but it is something to consider.

If you host this in your own daemon then you can control the AppDomain lifecycle, and could provide a clear suspend/resume model in your implementation of the Windows Service.

**Comment by dazinator** [comment]
Yeah I am thinking doing this in asp.net core so it could be hosted / run cross platform.

There are hooks in asp.net core to intercept startup and shutdown so could potentially handle this there:
 http://stackoverflow.com/questions/35257287/kestrel-shutdown-function-in-startup-cs-in-asp-net-core#35257670

SignalR in the browser will attempt to reconnect when the connection to the server is lost so I am guessing it would be possible to recover / survive from a server restart. In terns of where to store the business object though when this happens - thats a different question.. I'm not sure what serialiser should be used either

**Comment by wilfreds** [comment]
Your idea is not as crazy as you think. I have also been thinking about this. Keep the CSLA instance on the server. Sync it with javascript client object. And use your favorite javascript UI framework on the client.

Take a look at these cool projects done by Zang. He calls it WebBinding:
http://www.codeproject.com/Articles/747712/WebBinding-How-to-Bind-Client-JavaScript-Objects-t
http://www.codeproject.com/Articles/880922/Implement-WebBinding-using-AngularJS

In the knockout example 4 he demonstrates binding to object with children. In the example rows are added on the client.
I have done a simple test with binding to a CSLA object on the server. And it works. Even when the rows are added on the serverside. CSLA must be responsible for the factory of objects.

**Comment by tfreitasleal** [comment]
If I understand it correctly, the main goal is to have a desktop-like environment on the browser with real time (instantaneus) updates both sides.

In that case, I think is far simpler to use a development environment that does all that for you. Take a look at [Wisej](http://wisej.com/).

On [this thread](http://wisej.com/support/question/bound-dgv-not-updating?sort=oldest) I attached a sample of with static collection that updates the DataGridView on all browser instances.

> I run the sample on VS and it shows on Chrome. Then I open Firefox, IE and Edge, all running the sample (...).
>
> When I edit, add or remove a row on any of these browsers, it shows on the others at once, no need to refresh any browser.

This kind of behaviour is by design, no need to write a single line of Javascript. And this is a BIG BIG PLUS.

**Comment by dazinator** [comment]
@tfreitasleal - Thanks, I had a quick look at wisej and your code sample. I couldn't run the sample as I couldn't find the wisej assemblies - I'm assuming i have to sign up to get them as I couldn't see them on nuget.org either.

All in all, I don't think wisej is a good fit for me for a few reasons - all a matter of opinion. It's architecture seems to look quite invasive (although powerful). For example, my team uses MVC (razor, html, js) where as wisej basically looks like webforms but with a different base classes representing a "page", "form" etc etc. This isn't a good fit for us.

Also wisej doesn't seem to support netcore, and I am looking for a cross plat story. My goal would be that someone running an asp.net core website (on linux, windows, ios) should be able to use CSLA business objects to drive a real time MVC based web page, that uses whatever client side "data binding" javascript framework they are comfortable with for the client side MVVM (knockout, angular etc), but where this new fandangled library would:

1. Allow you to hand over a CSLA business object instance on the server to be used with the current page for real time updates.
2. Generate a client side "proxy" of the business object instance, that forwards any property updates to the server instance, and handles any property change notifications from the server to update the client side proxy property value.

Ultimately I'd be looking for a solution that:

1. Aligns well with MVC architecture and isn't too invasive.
2. Doesn't require purchasing a licence.
3. Allows me to use whichever data binding frameworks on the client, like KnockoutJs, Angular etc.
4. Exposes a proxy of the CSLA business object on the client (javascript object) that facilitates real time property updates with the server.

@wilfreds - that looks very promising, I'll look at those articles in more detail.

**Comment by tfreitasleal** [comment]
@dazinator 
You can download a [Wisej 30 days trial](http://wisej.com/#buy)

Wisej isn't WebForms but rather WinForms (same classes, same members, including DataBinding), It's very easy to work with: just C#  - no need for Javascript, CSS, HTML.
Of course I wouldn't advice anyone to use it in the context of an ASP.NET MVC project. Wisej environement can be used instead of ASP.NET XXX.

**Comment by dazinator** [comment]
Well guess what.. Blazor eventually came along ;-)
